hasher-matcher-actioner/hmalib/lambdas/api/middleware.py

```diff
@@ -24,27 +28,95 @@
 ...def get_all_datasets() -> CupcakesResponse:
 ...    response.status_code = 200 # optional
 ...    return CupcakesResponse(12, "key_lime")
+
+
+You can also specify the type of the request payload. If the request
+Content-Type is application/json and it can be parsed into the shape requested,
+your view will be called. Else, a 400 Bad request will be sent to the client.
+
+Sample usage:
+>>> @dataclass
+... class CupcakesRequest(DictParseable):
+...   @classmethod
+...   def from_dict(cls, d: t.Dict) -> 'CupcakesRequest':
+...     return CupcakesRequest(...)
+...
+>>> @app.post("/order-cupcakes/", apply=[jsoninator(CupcakesRequest)])
+... def order_cupcakes(request:CupcakesRequest) -> CupcakesResponse:
+...    <do stuff with request>
 """
 
+# This module is tested in ./tests/test_middleware.py. You can also refer to
+# that for sample usage.
+
 
 class JSONifiable:
     def to_json(self) -> t.Dict:
         raise NotImplementedError
 
 
-def jsoninator(view_fn: t.Callable[[int, int], JSONifiable]):
+class DictParseable:
+    @classmethod
+    def from_dict(cls, d: t.Dict) -> "DictParseable":
+        # Someday, include a vanilla de-serializer that uses dataclasses.fields
+        # to construct a dataclass on the fly.
+        raise NotImplementedError
+
+
+def jsoninator(
+    view_fn_or_request_type: t.Union[
+        t.Callable[[int, int], JSONifiable], t.Type[DictParseable]
+    ]
+):
     """
     Bottle plugin which allows you to create 'typed' views.
 
     Views are expected to return an object which has a to_json method. The
     content_type header will be automatically set, but if you need to set
     anything else, eg. status code or other headers, continue to use
     `bottle.response`.
+
+    If request type provided like jsoninator(RequestType), will de-serialize
+    request payload into the first argument to the view function.
     """
 
-    def wrapper(*args, **kwargs):
-        body = view_fn(*args, **kwargs)
-        response.content_type = "application/json"
-        return json.dumps(body.to_json())
+    # I feel like I'm doing a lot of work to support a common API for typed
+    # request payloads and untyped request payloads.  All this could be avoided
+    # by having `jsoninator` and `jsoninator_typed_request(RequestType)`, but
+    # meh! This looks shinier.
+
+    # Can't use isinstance because Class types are t.Callable as well.
+    if hasattr(view_fn_or_request_type, "from_dict"):
```

Comment:
I think it might be good use a `etmli5` comment or two here. decorators are hard enough to understand and this is a level beyond even if you understand python closures. Maybe little more details on `view_fn_or_request_type` and the check? My understanding of this is that an objectively worse but more actuate name would be `view_fn_OR_request_type_and_view_fn`
Something like if the request_type "from_dict"